CONTRIBUTING.md

```diff
@@ -0,0 +1,73 @@
+
+* The first line should ideally be limited to 50 characters, but must be no more than 70 characters.
+* The first line should not end with a full stop and the first word should be capitalised.
+* Use the imperative mood in the subject line - e.g. ```Fix a typo in CONTRIBUTING.md```, ```Remove if/else block in myfile.sh```, ```Eat your dinner```.
+* Insert a single blank line after the first line.
+* Subsequent lines should be wrapped at 72 characters.
+* Provide a detailed description of the change in the following lines, using the guidelines in the section below.
+* The last line of the commit should be a reference to the issue being fixed using the word JIRA, followed by the JIRA issue Key e.g. ```JIRA: HPCDIR-9999```
+
+In your commit message please consider the following points:
+
+* In the commit subject line, provide a brief description of the change.
+* The commit message must be descriptive and contain all the information required to fully understand and review the patch.
+* Do not assume the reviewer understands what the original problem was.
+* Do not assume the reviewer has access to external web services or site.
+* Do not assume the code is self-evident or self-documenting.
+* Describe why the change is being made.
+* Ensure sufficient information is provided to review your commit.
+* Each commit should ideally solve a single problem. Where there are several things to fix to reach the goal identified by the card, multiple commits should be used. Each commit should be as easy as possible to consume (or revert) as a single, complete change.
+
+### Pull requests
+
+* Pull Requests (PRs) should be made from a development branch in the main repo. 
+* A PR should contain a single commit, or multiple commits all related to the problem being solved by the PR.
+* The PR title and message should contain information relevant to why the commit(s) are happening. This can be based off the commit message.
+* As with commit messages, the PR description should reference the original issue.
+
+#### Reviewing pull requests
+
+When reviewing a PR, please ensure the included commit(s):
+
+* Actually works to fix the issue.
+* Passes any tests that are configured to run.
+* Contains a single commit, or multiple commits all related to the problem being solved by the PR.
+* Does not overreach. Each PR should be self contained to only address the issue at hand (see above).
+
+#### Merging pull requests
+
+In order for a PR to be merged, the following criteria should be met:
+
+* Any configured tests must have passed.
+* There must be at least 2 members of the engineering team review and give the PR a +1 approval using the review guidelines above.
+ * Use GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/) function so that your comments are collated into a single review.
+* The second +1 reviewer should merge the patch.
```

Comment:
Might be worth stating which merge mechanism to use if doing so manually.

Gitlab (and Github as well, I believe) use `git merge --no-ff <branch-name>`

I've worked at other places that prefer to use `git merge --squash <branch-name>`, while others just use  `git merge <branch-name>`

Reply:
Also, in platform dev, the author merges the code, as they are best placed to deal with any merge conflicts that arise. We should be consistent on this.

Do you have strong opinions on this?

Reply:
We have also generally asked the Author to merge the PR/MR and that is also how StackHPC generally works, as it is also the author that deploys the code to production.

Reply:
github actually now gives you an option in the gui to merge, squash and merge, or rebase and merge. On merging, I personally prefer to keep the granular commit history. I'd expect a dev to squash anything super trivial before creating the PR/MR, so instead of 47 commits a PR might only have 4 or 5 after squashing, but then whatever the dev has felt important enough to keep in separate commits should be retained on merge rather than squashing that down again.

The second point you mention about who merges the code is to prevent 5 people all looking, +1 it, and then leaving the merge to someone else. I'm not strongly opinionated - if author merge is the chosen mechanism I can update the doc

Reply:
`git merge --no-ff <branch-name>` is good because that keeps each commit, but keeps it contained so you can clearly see which branch it has come from. Also Gitlab does that by default, so I'd like us to use this if we can.

And yeah, I'd prefer author merges if pos, please